samcli/lib/test_runner/test_runner_template_generator.py

```diff
@@ -0,0 +1,347 @@
+from typing import *
+
+LOG = logging.getLogger(__name__)
+
+
+def __extract_api_id_from_arn(api_arn: str) -> str:
+    """
+    Extracts an api id from an HTTP or REST api arn.
+
+    NOTE: https://docs.aws.amazon.com/apigateway/latest/developerguide/arn-format-reference.html
+
+    REST API Arn: `arn:partition:apigateway:region::/restapis/api-id`
+
+    HTTP API Arn: `arn:partition:apigateway:region::/apis/api-id`
+
+    Parameters
+    ----------
+    api_arn : str
+        The arn of the HTTP or REST api from which the api id is extracted.
+
+    Returns
+    -------
+    str
+        The api id from the api arn.
+    """
+
+    return api_arn[api_arn.rindex("/") + 1 :]
+
+
+def __get_permissions_map() -> dict:
+    """
+    Returns a dictionary mapping a Regular Expression representing the arn of a specific resource to the default permissions generated for that resource.
+
+    Returns
+    -------
+    default_permissions_map : dict
+        A dictionary mapping a Regular Expression representing the arn of a resource to a list of IAM Action permissions generated for that resource.
+    """
+
+    partition_ex = "(aws|aws-cn|aws-us-gov)"
+    region_ex = "\w+-\w+-\d"
+    account_ex = "\d{12}"
+
+    default_permissions_map = {
+        # AWS::Lambda::Function
+        rf"^arn:{partition_ex}:lambda:{region_ex}:{account_ex}:function:[\w-]+(:\d+)?$": ["lambda:InvokeFunction"],
+        # AWS::ApiGateway::Api
+        rf"^arn:{partition_ex}:apigateway:{region_ex}::\/apis\/\w+$": ["execute-api:Invoke"],
+        # AWS::ApiGateway::RestApi
+        rf"^arn:{partition_ex}:apigateway:{region_ex}::\/restapis\/\w+$": ["execute-api:Invoke"],
+        # AWS::SQS::Queue
+        rf"^arn:{partition_ex}:sqs:{region_ex}:{account_ex}:[\w-]+$": ["sqs:SendMessage"],
+        # AWS::S3::Bucket
+        rf"^arn:{partition_ex}:s3:::[\w-]+$": ["s3:PutObject", "s3:GetObject"],
+        # AWS::Dynamodb::Table
+        rf"^arn:{partition_ex}:dynamodb:{region_ex}:{account_ex}:table\/[\w-]+$": [
+            "dynamodb:GetItem",
+            "dynamodb:PutItem",
+        ],
+        # AWS::StepFunctions::StateMachine
+        rf"^arn:{partition_ex}:states:{region_ex}:{account_ex}:stateMachine:[\w-]+$": [
+            "stepfunction:StartExecution",
+            "stepfunction:StopExecution",
+        ],
+    }
+    return default_permissions_map
+
+
+def __get_permissions(resource_arn: str) -> List[str]:
+    """
+    Returns a list of IAM Action permissions to generate given a resource arn.
+
+    An empty list is returned if there are no IAM Action permissions to generate for the given resource.
+
+    Parameters
+    ----------
+    resource_arn : str
+        The arn of the resource for which IAM Action permissions are returned.
+
+    Returns
+    -------
+    List[str]
+        The list of IAM Action permissions to generate for a given resource.
+    """
+
+    permissions_map = __get_permissions_map()
+
+    for arnExp in permissions_map.keys():
```

Comment:
You can iterate over `items()` instead, so you don't need the extra access when `return`ing.